File: backend/app/calibration/dates.py
from __future__ import annotations

import math
from datetime import UTC, datetime, timedelta
from typing import Literal

from app.calibration.transform import CalibrationError

UNIX_EPOCH = datetime(1970, 1, 1, tzinfo=UTC)

_SECONDS_PER_DAY = 86400.0

_DATE_FORMATS = (
    "%Y/%m/%d %H:%M:%S",
    "%Y-%m-%d %H:%M:%S",
    "%Y/%m/%d %H:%M",
    "%Y-%m-%d %H:%M",
    "%Y/%m/%d",
    "%Y-%m-%d",
)


def parse_axis_token(text: str) -> tuple[float, Literal["number", "date"]]:
    raw = text.strip()
    if not raw:
        raise CalibrationError(
            "Invalid axis token",
            hint="Enter a number or a date like YYYY/MM/DD",
        )
    for fmt in _DATE_FORMATS:
        try:
            naive = datetime.strptime(raw, fmt)
        except ValueError:
            continue
        aware = naive.replace(tzinfo=UTC)
        days = (aware - UNIX_EPOCH).total_seconds() / _SECONDS_PER_DAY
        return days, "date"
    try:
        value = float(raw)
    except ValueError as exc:
        raise CalibrationError(
            "Invalid axis token",
            hint="Enter a number or a date like YYYY/MM/DD",
        ) from exc
    if not math.isfinite(value):
        raise CalibrationError(
            "Invalid axis token",
            hint="Enter a finite number or a date like YYYY/MM/DD",
        )
    return value, "number"


def format_unix_days(value: float, pattern: str) -> str:
    if not math.isfinite(value):
        raise CalibrationError(
            "Invalid unix-days value",
            hint="Unix days must be a finite number",
        )
    dt = UNIX_EPOCH + timedelta(days=value)
    replacements = (
        ("YYYY", f"{dt.year:04d}"),
        ("MM", f"{dt.month:02d}"),
        ("DD", f"{dt.day:02d}"),
        ("hh", f"{dt.hour:02d}"),
        ("mm", f"{dt.minute:02d}"),
        ("ss", f"{dt.second:02d}"),
    )
    out = pattern
    for token, repl in replacements:
        out = out.replace(token, repl)
    return out

File: frontend/src/lib/dates.ts
import type { Scale } from '../types'
import { CalibrationError } from './transform2d'

const MS_PER_DAY = 86400 * 1000

const DATE_RE =
  /^(\d{4})([/-])(\d{1,2})\2(\d{1,2})(?: (\d{1,2}):(\d{2})(?::(\d{2}))?)?$/

const NUMBER_RE = /^[+-]?(?:\d+\.?\d*|\.\d+)(?:[eE][+-]?\d+)?$/

export type AxisTokenKind = 'number' | 'date'

function invalidToken(): never {
  throw new CalibrationError(
    'Invalid axis token',
    'Enter a number or a date like YYYY/MM/DD',
  )
}

export function parseAxisToken(text: string): [number, AxisTokenKind] {
  const raw = text.trim()
  if (!raw) invalidToken()

  const dateMatch = DATE_RE.exec(raw)
  if (dateMatch) {
    const year = Number(dateMatch[1])
    const month = Number(dateMatch[3])
    const day = Number(dateMatch[4])
    const hour = dateMatch[5] !== undefined ? Number(dateMatch[5]) : 0
    const minute = dateMatch[6] !== undefined ? Number(dateMatch[6]) : 0
    const second = dateMatch[7] !== undefined ? Number(dateMatch[7]) : 0
    if (month < 1 || month > 12 || day < 1 || hour > 23 || minute > 59 || second > 59) {
      invalidToken()
    }
    const ms = Date.UTC(year, month - 1, day, hour, minute, second)
    const check = new Date(ms)
    if (
      check.getUTCFullYear() !== year ||
      check.getUTCMonth() !== month - 1 ||
      check.getUTCDate() !== day ||
      check.getUTCHours() !== hour ||
      check.getUTCMinutes() !== minute ||
      check.getUTCSeconds() !== second
    ) {
      invalidToken()
    }
    return [ms / MS_PER_DAY, 'date']
  }

  if (!NUMBER_RE.test(raw)) invalidToken()
  const value = Number(raw)
  if (!Number.isFinite(value)) {
    throw new CalibrationError(
      'Invalid axis token',
      'Enter a finite number or a date like YYYY/MM/DD',
    )
  }
  return [value, 'number']
}

export function formatUnixDays(value: number, pattern: string): string {
  if (!Number.isFinite(value)) {
    throw new CalibrationError('Invalid unix-days value', 'Unix days must be a finite number')
  }
  const dt = new Date(value * MS_PER_DAY)
  const replacements: Array<[string, string]> = [
    ['YYYY', String(dt.getUTCFullYear()).padStart(4, '0')],
    ['MM', String(dt.getUTCMonth() + 1).padStart(2, '0')],
    ['DD', String(dt.getUTCDate()).padStart(2, '0')],
    ['hh', String(dt.getUTCHours()).padStart(2, '0')],
    ['mm', String(dt.getUTCMinutes()).padStart(2, '0')],
    ['ss', String(dt.getUTCSeconds()).padStart(2, '0')],
  ]
  let out = pattern
  for (const [token, repl] of replacements) {
    out = out.split(token).join(repl)
  }
  return out
}

export function formatBoundValue(value: number | null, scale: Scale): string {
  if (value == null) return ''
  if (scale === 'date') {
    const pattern = Math.abs(value - Math.round(value)) > 1e-6 ? 'YYYY/MM/DD hh:mm:ss' : 'YYYY/MM/DD'
    return formatUnixDays(value, pattern)
  }
  return String(value)
}

export function parseBoundValue(raw: string, scale: Scale): number {
  const [value, kind] = parseAxisToken(raw)
  if (scale === 'log' && kind === 'date') {
    throw new CalibrationError(
      'Log and date cannot be used on the same axis',
      'Choose either log or date',
    )
  }
  if (scale === 'log' && value <= 0) {
    throw new CalibrationError(
      'Log scale requires values > 0',
      'Enter a value greater than zero',
    )
  }
  return value
}
